### ISSUE TITLE
Debian packaging and some issues

### ISSUE THREAD

**Issue by charles2910** [MasterQ32/kristall#156]
Hi, I've come across gemini and realized that there aren't apps for accessing the protocol in the Debian repositories. I've tried a few of them and really liked kristall, so I decided to package it (thanks for kristall BTW).

A couple of things popped up (mostly minor problems but one really big). So the little things first. Lintian has pointed out some typos:
```
I: kristall: hardening-no-fortify-functions usr/bin/kristall
I: kristall: spelling-error-in-binary usr/bin/kristall "allows to" "allows one to"
I: kristall: spelling-error-in-binary usr/bin/kristall "amount of times" "number of times"
I: kristall: spelling-error-in-binary usr/bin/kristall accidentially accidentally
I: kristall: spelling-error-in-binary usr/bin/kristall completly completely
```
I've fixed them with the patch below.

<details>
<summary>040-fix_typos.patch</summary>

```
Description: fixes typos in the source code
Author: Carlos Henrique Lima Melara <charlesmelara@outlook.com>
Forwarded: https://github.com/MasterQ32/kristall/issues/148
Last-Update: 2021-02-17

--- kristall-0.3+git20210209.0843ee2.orig/src/about/help.gemini
+++ kristall-0.3+git20210209.0843ee2/src/about/help.gemini
@@ -92,7 +92,7 @@ This tab contains an unsorted list of se
 
 [Enabled Protocols] allows you to fine-tune which protocols are fetched by Kristall. By default, only Gemini is enabled, all other protocols are disabled. Disabled protocols are either not served with an error message or forwarded to your OS handler for that URL scheme.
 
-[Text Rendering] allows to control whether Kristall parses text input files or not. This is usually set to [Fancy] which renders text/html, text/gemini, text/markdown and text/gophermap to a nice, hyperlinked display. When set to [Always plain text], Kristall will display all text/* files as plaintext files instead. This may be inconvenient, but necessary for misparsed sites.
+[Text Rendering] allows one to control whether Kristall parses text input files or not. This is usually set to [Fancy] which renders text/html, text/gemini, text/markdown and text/gophermap to a nice, hyperlinked display. When set to [Always plain text], Kristall will display all text/* files as plaintext files instead. This may be inconvenient, but necessary for misparsed sites.
 
 [Enable text highlights] allows you to enable an experimental feature that allows *highlighting* and _underlining_ text in text/gemini documents. It will probably misrender, but you can try it out anyways.
 
@@ -100,7 +100,7 @@ This tab contains an unsorted list of se
 
 [Unknown Scheme] changes the behaviour how Kristall handles unknown/disabled URL schemes. [Use OS default handler] will invoke your OS default, [Display error message] will just pop up a message box and tell you that Kristall cannot handle this URL.
 
-[Max. Number of Redirections] is a setting that allows you to restrict sites to redirect you only a certain amount of times before erroring out. Setting this to 0 will disable redirections completly, displaying an error with the target URL.
+[Max. Number of Redirections] is a setting that allows you to restrict sites to redirect you only a certain number of times before erroring out. Setting this to 0 will disable redirections completely, displaying an error with the target URL.
 
 [Redirection Handling] allows you to fine-tune the way Kristall allows redirections. Each of the options defines if Kristall should ask you to allow the redirect or do it silently. [Ask for cross-scheme redirection] will pop up a message box if a host tries to redirect you from one URL scheme to another, e.g. when a web server redirects you from HTTP to HTTPS. [Ask for cross-host redirection] will pop up the message box for all redirections through host boundaries, e.g. when example.com redirects you to www.example.com. [Ask for cross-scheme or cross-host redirection] will enable both of the previous behaviours, asking when any cross-boundary redirection happens. [Ask for all redirections] will pop up a message box every time a server tries to redirect you, keeping you in full control over all redirections. [Silently redirect everything] is the exact oppositve of that, accepting all redirections without warning or notice.
 
@@ -145,7 +145,7 @@ The lone text with with the [host.name]
 ### Gemini TLS and HTTPS TLS
 These two sites contain the TLS settings for either Gemini or HTTPS. Both protocols are handled in the same way, but with different data sets, so each one has its own settings page.
 
-[Trust Level] defines how you trust hosts. [Trust on first encounter] is also known as *Trust On First Use* (or TOFU) and will store the servers public key in Kristalls database of trusted hosts. If a host is later encountered that has changed its public key, an error will be displayed to the user that this host may be compromised (as the changing of a public key can be a man-in-the-middle attack). [Trust everything] will just happily accept every TLS server, ignoring the certificate issuer completly. [Manually verify fingerprints] allows you to chose whether you trust a server or not based on its fingerprint. This will be displayed in the error page as well as the option to add that server to your list of trusted hosts.
+[Trust Level] defines how you trust hosts. [Trust on first encounter] is also known as *Trust On First Use* (or TOFU) and will store the servers public key in Kristalls database of trusted hosts. If a host is later encountered that has changed its public key, an error will be displayed to the user that this host may be compromised (as the changing of a public key can be a man-in-the-middle attack). [Trust everything] will just happily accept every TLS server, ignoring the certificate issuer completely. [Manually verify fingerprints] allows you to chose whether you trust a server or not based on its fingerprint. This will be displayed in the error page as well as the option to add that server to your list of trusted hosts.
 
 [Certificate Authorities] allows you to enable/disable the use of your systems CA trust store. Sites that can be enabled via the CA system will not be added to the list of trusted hosts (as it is only meant for TOFU/manual implementation), but will not error out. [Use local certificate authorities] will enable that behaviour, [Don't use local certificate authorities] will disable it.
 
@@ -166,7 +166,7 @@ When selecting a certificate, its detail
 [Common Name] is the CN value that was used when creating the certificate. Its used as a identifier and the only required field when creating the cert. You cannot change this.
 [Expiration Date] is the date when your certificate expires.
 [Expires in] shows the numbe of days until your certificate expires. This may be more intuitive to work with, but communicating the expiration date is recommended.
-[Host Filter] is a security-measurement to shield you from accidental identity exposure. You can type in a URL with wildcards, using ? for a single character, * for any number of characters, including zero and […] for allowing a set of certain characters to be matched. When you try activating the certificate on a URL that does not match your Host Filter, Kristall will ask you if you really want to enable the certificate. This prevents you from accidentially using the certificate on a host or URL where it shouldn't be used.
+[Host Filter] is a security-measurement to shield you from accidental identity exposure. You can type in a URL with wildcards, using ? for a single character, * for any number of characters, including zero and […] for allowing a set of certain characters to be matched. When you try activating the certificate on a URL that does not match your Host Filter, Kristall will ask you if you really want to enable the certificate. This prevents you from accidentally using the certificate on a host or URL where it shouldn't be used.
 [Auto-Enable Certificate] is built on top of the Host Filter. When you don't have a client certificate enabled *and* you visit a URL that matches the Host Filter property, Kristall will ask you if you want to enable that certificate. This is convenient when you need a certificate to visit that location anyways and this allows you to quickly enable your default certificate.
 [Fingerprint] is the SHA256 fingerprint of this certificate.
 [Notes] is a free-form text field for your private use. Kristall does not use this value what-so-ever. Use this field to make notes about that certificate.
```

</details>

Another thing happened  while checking the hardening of the build. Debian attach some build flags to the build env, but the qmake wasn't receiving it. It was resolved by the patch below (not sure if you'd like to add to the kristall source).

<details>
<summary>030-Add_hardening_flags.patch</summary>

```
Description: adds buildflags to qmake
Author: Carlos Henrique Lima Melara <charlesmelara@outlook.com>
Forwarded: https://github.com/MasterQ32/kristall/issues/148
Last-Update: 2021-02-17

--- kristall-0.3+git20210209.0843ee2.orig/src/kristall.pro
+++ kristall-0.3+git20210209.0843ee2/src/kristall.pro
@@ -21,6 +21,11 @@ DEFINES += KRISTALL_VERSION="\"$(shell c
 # We need iconv on non-linux platforms
 !linux: LIBS += -liconv
 
+QMAKE_CPPFLAGS *= $(shell dpkg-buildflags --get CPPFLAGS)
+QMAKE_CFLAGS   *= $(shell dpkg-buildflags --get CFLAGS)
+QMAKE_CXXFLAGS *= $(shell dpkg-buildflags --get CXXFLAGS)
+QMAKE_LFLAGS   *= $(shell dpkg-buildflags --get LDFLAGS)
+
 QMAKE_CFLAGS += -Wno-unused-parameter -Werror=return-type
 QMAKE_CXXFLAGS += -Wno-unused-parameter -Werror=return-type
```

</details>

Now to the big problem. Lintian has caused the following `E: kristall: possible-gpl-code-linked-with-openssl`. And the problem is an incompatibility between the openSSL license and the GPL-3 (it's better explained [here](https://people.gnome.org/~markmc/openssl-and-the-gpl.html)). There are a few solutions:

1. Add an exemption to the GPL-3 to link against the openssl lib and add the openssl license to the source code (as LICENSE.openssl for example), the process is better detailed [here](https://lists.debian.org/debian-legal/2015/09/msg00000.html);
2. Link against another cripto libs such as gnuTLS or libNSS or other;
3. Wait for the new version of the openssl lib 3.0 that changed the license to apache 2.0 which is compatible with the gpl-3, although I'm not sure when it'll be released.

I can help with any questions on these problems and if you'd like a PR instead of the patches, let me know.

Cheers,
Charles

**Comment by MasterQ32** [comment]
Heya! Can you please make separate PRs? I'd be happy to merge the typo patch without much looking. for the hardening: it seems very dpkg-specific and will probably require some work to allow building on systems without dpkg/apt.

For the licencing thing: Option 1 seems like the best thing right now, or even better: get rid of OpenSSL completly (i only use it for generating the certificates, i think there are other options for that)

**Comment by charles2910** [comment]
Hey, thanks for the quick answer.

I'll do the PRs in the weekend.

You only use the openssl in `src/certificatehelper.cpp`, right? It may be easier to just switch to gnutls library (in a similar [bug](https://lists.debian.org/debian-legal/2015/09/msg00005.html) the folks just switched). If that is the only file, I'll look into it the weekend too.

**Comment by charles2910** [comment]
Hi, not exactly in the weekend XD but here we go.

I've made the changes and the PR's (#170 and #171).

Also found some good news in the weekend. Turns out that last year Debian started to consider openssl a system library which resolves the conflict of licenses, so it's not needed to add the license exemption. (The problem is that they haven't released the new version of lintian with this change)
